File: v4/rust/iprange-livedb/src/publication/file_inspection.rs
//! Stable, bounded inspection of publication output bytes.

use std::fs::File;

use crate::bootstrap::{self, OpenMode};
use crate::cancellation::CancellationToken;
use crate::contract::MetaV4;
use crate::error::ErrorCode;
use crate::live_lock::{self, Mode};
use crate::live_sidecar::MAIN_LIFETIME_LOCK;
use crate::mapping::Mapping;

use super::namespace::{Destination, Identity, Name, Regular};
use super::output;
use super::problem::Problem;
use super::reservation::Header;
use super::result::AccessPolicy;
use super::{ArtifactKind, DirectoryRole};

#[derive(Clone, Copy, Debug, PartialEq, Eq)]
pub(super) enum Content {
    Desired,
    Other,
}

#[derive(Clone, Copy, Debug, PartialEq, Eq)]
enum Location {
    Main,
    Private,
}

#[derive(Debug)]
pub(super) struct Inspected {
    pub(super) name: Name,
    pub(super) file: File,
    pub(super) mapping: Mapping,
    pub(super) identity: Identity,
    pub(super) meta: MetaV4,
    pub(super) byte_length: u64,
    pub(super) sha512: [u8; 64],
    pub(super) content: Content,
    pub(super) access: AccessPolicy,
    location: Location,
    attempt_id: [u8; 16],
}

impl Inspected {
    pub(super) fn verify(&self, destination: &Destination) -> Result<(), Problem> {
        require_available(
            destination,
            self.location,
            self.attempt_id,
            &self.name,
            self.identity,
        )?;
        let (meta, byte_length) = read_bootstrap(&self.file, &self.mapping)?;
        if (meta, byte_length) != (self.meta, self.byte_length) {
            return Err(conflict("publication output changed after inspection"));
        }
        destination
            .directory()
            .verify()
            .and_then(|()| {
                destination
                    .directory()
                    .verify_name(&self.name, self.identity)
            })
            .map_err(|error| Problem::namespace(&error))
    }
}

pub(super) fn main(
    destination: &Destination,
    header: Header,
    cancellation: &CancellationToken,
) -> Result<Option<Inspected>, Problem> {
    #[cfg(target_os = "freebsd")]
    let Some(regular) = destination
        .directory()
        .open_regular_any_link(destination.main(), false)
        .map_err(|error| Problem::namespace(&error))?
    else {
        return Ok(None);
    };
    #[cfg(not(target_os = "freebsd"))]
    let Some(regular) = destination
        .directory()
        .open_regular(destination.main(), false)
        .map_err(|error| Problem::namespace(&error))?
    else {
        return Ok(None);
    };
    #[cfg(target_os = "freebsd")]
    if regular.identity.encode() == header.output_identity {
        let private = destination
            .output_name(header.attempt_id)
            .map_err(|error| Problem::namespace(&error))?;
        destination
            .directory()
            .finish_noreplace_transition(&private, destination.main(), regular.identity)
            .map_err(|error| Problem::namespace(&error))?;
    }
    inspect(
        destination,
        destination.main().clone(),
        regular,
        header,
        Location::Main,
        Mode::Shared,
        cancellation,
    )
    .map(Some)
}

pub(super) fn private(
    destination: &Destination,
    header: Header,
    cancellation: &CancellationToken,
) -> Result<Option<Inspected>, Problem> {
    let Some(inspected) = private_owned(destination, header, cancellation)? else {
        return Ok(None);
    };
    if inspected.access != AccessPolicy::CreatorOnly {
        return Err(conflict(
            "private publication output access no longer matches its reservation",
        ));
    }
    Ok(Some(inspected))
}

pub(super) fn private_owned(
    destination: &Destination,
    header: Header,
    cancellation: &CancellationToken,
) -> Result<Option<Inspected>, Problem> {
    let name = destination
        .output_name(header.attempt_id)
        .map_err(|error| Problem::namespace(&error))?;
    let Some(regular) = destination
        .directory()
        .open_regular(&name, true)
        .map_err(|error| Problem::namespace(&error))?
    else {
        return Ok(None);
    };
    let inspected = inspect(
        destination,
        name,
        regular,
        header,
        Location::Private,
        Mode::Exclusive,
        cancellation,
    )?;
    require_exact_private(&inspected, header)?;
    Ok(Some(inspected))
}

fn inspect(
    destination: &Destination,
    name: Name,
    regular: Regular,
    header: Header,
    location: Location,
    mode: Mode,
    cancellation: &CancellationToken,
) -> Result<Inspected, Problem> {
    cancellation.check().map_err(|error| Problem::sdk(&error))?;
    require_available(
        destination,
        location,
        header.attempt_id,
        &name,
        regular.identity,
    )?;
    lock_output(&regular, mode, cancellation)?;
    destination
        .directory()
        .verify_name(&name, regular.identity)
        .map_err(|error| Problem::namespace(&error))?;
    let (mapping, meta, byte_length) = map_bootstrap(&regular.file)?;
    let sha512 = output::digest_cancellable(&mapping, byte_length, cancellation)
        .map_err(|error| Problem::output(&error))?;
    let (final_meta, final_length) = read_bootstrap(&regular.file, &mapping)?;
    if (final_meta, final_length) != (meta, byte_length) {
        return Err(conflict("publication output changed while hashing"));
    }
    destination
        .directory()
        .verify()
        .and_then(|()| destination.directory().verify_name(&name, regular.identity))
        .map_err(|error| Problem::namespace(&error))?;
    let content = classify(meta, byte_length, sha512, header);
    let access = classify_access(&regular, header);
    Ok(Inspected {
        name,
        file: regular.file,
        mapping,
        identity: regular.identity,
        meta,
        byte_length,
        sha512,
        content,
        access,
        location,
        attempt_id: header.attempt_id,
    })
}

fn require_available(
    destination: &Destination,
    location: Location,
    attempt_id: [u8; 16],
    name: &Name,
    identity: Identity,
) -> Result<(), Problem> {
    if location == Location::Private {
        super::gc_barrier::require_source_available(
            destination.directory(),
            attempt_id,
            0,
            ArtifactKind::PrivateOutput,
            DirectoryRole::Destination,
            name,
            identity,
        )?;
    }
    Ok(())
}

fn lock_output(
    regular: &Regular,
    mode: Mode,
    cancellation: &CancellationToken,
) -> Result<(), Problem> {
    live_lock::lock_file_cancellable(&regular.file, MAIN_LIFETIME_LOCK, mode, cancellation)
        .map_err(|error| Problem::sdk(&error))?;
    cancellation.check().map_err(|error| Problem::sdk(&error))
}

fn map_bootstrap(file: &File) -> Result<(Mapping, MetaV4, u64), Problem> {
    let byte_length = file
        .metadata()
        .map_err(crate::error::Error::from)
        .map_err(|error| Problem::sdk(&error))?
        .len();
    if !bootstrap::geometry_valid(byte_length) {
        return Err(conflict(
            "publication destination has invalid v4 file geometry",
        ));
    }
    let mapping =
        Mapping::read_only_view(file, byte_length).map_err(|error| Problem::sdk(&error))?;
    let (meta, _) = read_bootstrap(file, &mapping)?;
    Ok((mapping, meta, byte_length))
}

fn read_bootstrap(file: &File, mapping: &Mapping) -> Result<(MetaV4, u64), Problem> {
    let byte_length = file
        .metadata()
        .map_err(crate::error::Error::from)
        .map_err(|error| Problem::sdk(&error))?
        .len();
    if byte_length != mapping.len() {
        return Err(conflict(
            "publication destination changed while reading metadata",
        ));
    }
    let page0 = mapping.page(0, 2).map_err(|error| Problem::sdk(&error))?;
    let page1 = mapping.page(1, 2).map_err(|error| Problem::sdk(&error))?;
    let opened = bootstrap::open_meta_pages(page0, page1, byte_length, OpenMode::ImmutableReader)
        .map_err(|_| conflict("publication destination is not a complete v4 file"))?;
    Ok((opened.meta, byte_length))
}

fn classify(meta: MetaV4, byte_length: u64, sha512: [u8; 64], header: Header) -> Content {
    if meta.database_id == header.database_id
        && meta.txn_id == header.transaction_id
        && meta.commit_nonce == header.commit_nonce
        && byte_length == header.output_byte_length
        && sha512 == header.output_sha512
    {
        Content::Desired
    } else {
        Content::Other
    }
}

fn classify_access(regular: &Regular, header: Header) -> AccessPolicy {
    match regular.creator_only_commitment() {
        Ok(commitment) if commitment == header.security_commitment => AccessPolicy::CreatorOnly,
        _ => AccessPolicy::ChangedOrUnproven,
    }
}

fn require_exact_private(inspected: &Inspected, header: Header) -> Result<(), Problem> {
    if inspected.identity.encode() != header.output_identity
        || inspected.content != Content::Desired
    {
        return Err(conflict(
            "private publication output does not match its reservation",
        ));
    }
    Ok(())
}

const fn conflict(detail: &'static str) -> Problem {
    Problem::new(ErrorCode::Conflict, None, detail)
}

File: v4/go/internal/publication/file_inspection.go
// Stable, bounded inspection of publication output bytes (Rust
// publication/file_inspection.rs). The resolver inspects the main and
// private outputs without validating the whole file: the meta pair,
// byte length, and SHA-512 are compared with the reservation record,
// the output lifetime lock is held for the inspection, and custody is
// re-proved at the exact names. The gc-barrier availability call runs
// at the private position (windows arm in gc_barrier_windows.go).

package publication

import (
	"os"

	"github.com/firehol/iprange/v4/go/internal/bootstrap"
	"github.com/firehol/iprange/v4/go/internal/format"
	"github.com/firehol/iprange/v4/go/internal/live"
	"github.com/firehol/iprange/v4/go/internal/mapping"
	"github.com/firehol/iprange/v4/go/internal/security"
)

// outputContent classifies one inspected output against the
// reservation header (Rust file_inspection::Content).
type outputContent uint8

const (
	outputContentDesired outputContent = iota
	outputContentOther
)

// inspectedOutput is one inspected publication output with its
// lifetime lock held (Rust file_inspection::Inspected). Close
// releases the mapped view and the descriptor exactly where Rust
// drops the value.
type inspectedOutput struct {
	name       string
	file       *os.File
	mapping    *mapping.Mapping
	identity   live.FileIdentity
	meta       format.Meta
	byteLength uint64
	sha512     [64]byte
	content    outputContent
	access     AccessPolicy
	location   outputLocation
	attemptID  [16]byte
}

// Close releases the inspected output resources (Rust drop of
// Inspected; the mapping close also releases its duplicated
// descriptor).
func (o *inspectedOutput) Close() error {
	var first error
	if o.mapping != nil {
		if err := o.mapping.Close(); err != nil && first == nil {
			first = err
		}
	}
	if o.file != nil {
		if err := o.file.Close(); err != nil && first == nil {
			first = err
		}
	}
	return first
}

// verify re-proves one inspected output exactly as inspected (Rust
// Inspected::verify): the geo facts must be unchanged and the name
// must still carry the retained inode after a directory proof.
func (o *inspectedOutput) verify(destination *destination) error {
	meta, byteLength, err := readBootstrap(o.file, o.mapping)
	if err != nil {
		return err
	}
	if meta != o.meta || byteLength != o.byteLength {
		return conflictProblem("publication output changed after inspection")
	}
	if err := destination.directory().Verify(); err != nil {
		return namespaceProblem(err)
	}
	if err := destination.directory().VerifyName(o.name, o.identity); err != nil {
		return namespaceProblem(err)
	}
	return nil
}

// inspectMainOutput inspects the main destination output (Rust
// file_inspection::main): the lifetime lock is shared because the
// main may be live-published by a concurrent writer.
func inspectMainOutput(destination *destination, header reservationHeader, check func() error) (*inspectedOutput, error) {
	regular, err := openMainRegular(destination, header)
	if err != nil {
		return nil, err
	}
	if regular == nil {
		return nil, nil
	}
	inspected, err := inspectOutput(destination, destination.mainName(), regular, header, outputLocationMain, false, check)
	if err != nil {
		_ = regular.File.Close()
		return nil, err
	}
	return inspected, nil
}

// inspectPrivateOutput opens the private output and refuses it
// unless it exactly matches the reservation record (Rust
// file_inspection::private_owned: require_exact_private; the
// resolved-main and abandon cleanups only discard an exact private
// output).
func inspectPrivateOutput(destination *destination, header reservationHeader, check func() error) (*inspectedOutput, error) {
	inspected, err := inspectPrivateOutputUnchecked(destination, header, check)
	if err != nil {
		return nil, err
	}
	if inspected == nil {
		return nil, nil
	}
	if err := requireExactPrivate(inspected, header); err != nil {
		_ = inspected.Close()
		return nil, err
	}
	return inspected, nil
}

// inspectPrivateOutputExact opens the private output and refuses it
// unless it is creator-only and exactly matches the reservation
// (Rust file_inspection::private: the completion path adds the
// creator-only access proof).
func inspectPrivateOutputExact(destination *destination, header reservationHeader, check func() error) (*inspectedOutput, error) {
	inspected, err := inspectPrivateOutput(destination, header, check)
	if err != nil {
		return nil, err
	}
	if inspected == nil {
		return nil, nil
	}
	if inspected.access != AccessPolicyCreatorOnly {
		_ = inspected.Close()
		return nil, conflictProblem("private publication output access no longer matches its reservation")
	}
	return inspected, nil
}

// inspectPrivateOutputUnchecked runs the exact inspection machine
// over the private output without the exact-record refusal (Rust
// private_owned without require_exact_private).
func inspectPrivateOutputUnchecked(destination *destination, header reservationHeader, check func() error) (*inspectedOutput, error) {
	name, err := destination.outputName(header.attemptID)
	if err != nil {
		return nil, err
	}
	regular, err := destination.directory().OpenRegular(name, true)
	if err != nil {
		return nil, err
	}
	if regular == nil {
		return nil, nil
	}
	inspected, err := inspectOutput(destination, name, regular, header, outputLocationPrivate, true, check)
	if err != nil {
		_ = regular.File.Close()
		return nil, err
	}
	return inspected, nil
}

// requireExactPrivate refuses one inspected private output whose
// inode or content no longer matches the reservation (Rust
// require_exact_private).
func requireExactPrivate(inspected *inspectedOutput, header reservationHeader) error {
	if reservationIdentityBytes(inspected.identity) != header.outputIdentity ||
		inspected.content != outputContentDesired {
		return conflictProblem("private publication output does not match its reservation")
	}
	return nil
}

// inspectOutput runs the exact inspection machine over one open
// output (Rust file_inspection::inspect; the gc-barrier availability
// call runs at the private position before the output lock, exactly
// like the Rust require_available).
func inspectOutput(destination *destination, name string, regular *live.RegularFile, header reservationHeader, location outputLocation, exclusive bool, check func() error) (*inspectedOutput, error) {
	if err := checkCancellation(check); err != nil {
		return nil, err
	}
	if location == outputLocationPrivate {
		if err := requireSourceAvailable(destination.directory(), header.attemptID, 0, ArtifactPrivateOutput, DirectoryRoleDestination, name, regular.Identity); err != nil {
			return nil, err
		}
	}
	if err := lockOutputFile(regular.File, exclusive, check); err != nil {
		return nil, err
	}
	if err := destination.directory().VerifyName(name, regular.Identity); err != nil {
		return nil, err
	}
	mapped, meta, byteLength, err := mapBootstrap(regular.File)
	if err != nil {
		return nil, err
	}
	sha512, err := digestCancellable(mapped, byteLength, check)
	if err != nil {
		_ = mapped.Close()
		return nil, outputProblem(err)
	}
	finalMeta, finalLength, err := readBootstrap(regular.File, mapped)
	if err != nil {
		_ = mapped.Close()
		return nil, err
	}
	if finalMeta != meta || finalLength != byteLength {
		_ = mapped.Close()
		return nil, conflictProblem("publication output changed while hashing")
	}
	if err := destination.directory().Verify(); err != nil {
		_ = mapped.Close()
		return nil, err
	}
	if err := destination.directory().VerifyName(name, regular.Identity); err != nil {
		_ = mapped.Close()
		return nil, err
	}
	return &inspectedOutput{
		name:       name,
		file:       regular.File,
		mapping:    mapped,
		identity:   regular.Identity,
		meta:       meta,
		byteLength: byteLength,
		sha512:     sha512,
		content:    classifyOutput(meta, byteLength, sha512, header),
		access:     classifyOutputAccess(regular, header),
		location:   location,
		attemptID:  header.attemptID,
	}, nil
}

// lockOutputFile takes the output lifetime lock cancellably (Rust
// file_inspection::lock_output; main outputs lock shared, the exact
// private output locks exclusive).
func lockOutputFile(file *os.File, exclusive bool, check func() error) error {
	mode := live.LockShared
	if exclusive {
		mode = live.LockExclusive
	}
	if err := live.LockFileCancellable(file, live.MainLifetimeOffset, mode, check); err != nil {
		return err
	}
	return checkCancellation(check)
}

// mapBootstrap maps the output read-only and reads its meta pair
// (Rust file_inspection::map_bootstrap: the geometry must be valid
// for the v4 format before the mapping is created).
func mapBootstrap(file *os.File) (*mapping.Mapping, format.Meta, uint64, error) {
	byteLength, err := fstatSize(file)
	if err != nil {
		return nil, format.Meta{}, 0, err
	}
	if !reservationGeometryValid(byteLength) {
		return nil, format.Meta{}, 0, conflictProblem("publication destination has invalid v4 file geometry")
	}
	mapped, err := mapping.MapFile(file, byteLength, false)
	if err != nil {
		return nil, format.Meta{}, 0, err
	}
	meta, _, err := readBootstrap(file, mapped)
	if err != nil {
		_ = mapped.Close()
		return nil, format.Meta{}, 0, err
	}
	return mapped, meta, byteLength, nil
}

// readBootstrap re-reads the meta pair and physical length and proves
// nothing changed (Rust file_inspection::read_bootstrap).
func readBootstrap(file *os.File, mapped *mapping.Mapping) (format.Meta, uint64, error) {
	byteLength, err := fstatSize(file)
	if err != nil {
		return format.Meta{}, 0, sdkProblem(err)
	}
	if byteLength != mapped.Size() {
		return format.Meta{}, 0, conflictProblem("publication destination changed while reading metadata")
	}
	page0, err := mapped.Page(0)
	if err != nil {
		return format.Meta{}, 0, sdkProblem(err)
	}
	page1, err := mapped.Page(1)
	if err != nil {
		return format.Meta{}, 0, sdkProblem(err)
	}
	meta, err := bootstrap.OpenMeta(page0, page1, byteLength, bootstrap.ModeImmutableReader)
	if err != nil {
		return format.Meta{}, 0, conflictProblem("publication destination is not a complete v4 file")
	}
	return meta, byteLength, nil
}

// classifyOutput compares one inspected output with the reservation
// header facts (Rust file_inspection::classify).
func classifyOutput(meta format.Meta, byteLength uint64, sha512 [64]byte, header reservationHeader) outputContent {
	if meta.DatabaseID == header.databaseID &&
		meta.TxnID == header.transactionID &&
		meta.CommitNonce == header.commitNonce &&
		byteLength == header.outputByteLength &&
		sha512 == header.outputSHA512 {
		return outputContentDesired
	}
	return outputContentOther
}

// classifyOutputAccess derives the creator-only policy of one
// inspected output (Rust file_inspection::classify_access).
func classifyOutputAccess(regular *live.RegularFile, header reservationHeader) AccessPolicy {
	commitment, err := security.CreatorOnlyCommitment(regular.File)
	if err == nil && commitment == header.securityCommitment {
		return AccessPolicyCreatorOnly
	}
	return AccessPolicyChangedOrUnproven
}
